File: src/org/cappuccino_project/ide/intellij/plugin/psi/utils/ObjJTreeUtil.java
package org.cappuccino_project.ide.intellij.plugin.psi.utils;

import com.intellij.formatting.Spacing;
import com.intellij.lang.ASTNode;
import com.intellij.psi.*;
import com.intellij.psi.impl.source.tree.TreeUtil;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.util.PsiTreeUtil;
import org.cappuccino_project.ide.intellij.plugin.formatting.ObjJSpacingBuilder;
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJComment;
import org.cappuccino_project.ide.intellij.plugin.psi.types.ObjJTypes;
import org.cappuccino_project.ide.intellij.plugin.utils.ArrayUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.logging.Logger;

import static org.cappuccino_project.ide.intellij.plugin.psi.types.TokenSets.*;

public class ObjJTreeUtil extends PsiTreeUtil{

    private static final Logger LOGGER = Logger.getLogger(ObjJTreeUtil.class.getName());

    @NotNull
    public static <StubT extends StubElement>  List<StubT> filterStubChildren(StubElement<com.intellij.psi.PsiElement> parent, Class<StubT> stubClass) {
        if (parent == null) {
            return Collections.emptyList();
        }
        return filterStubChildren(parent.getChildrenStubs(), stubClass);
    }

    @NotNull
    public static <StubT extends StubElement>  List<StubT> filterStubChildren(List<StubElement> children, Class<StubT> stubClass) {
        if (children == null) {
            return Collections.emptyList();
        }
        return ArrayUtils.filter(children, stubClass);
    }

    @NotNull
    public static List<PsiElement> getChildrenOfType(@NotNull PsiElement element, @NotNull IElementType iElementType) {
        List<PsiElement> out = new ArrayList<>();
        for (PsiElement child : element.getChildren()) {
            //LOGGER.log(Level.INFO, "Child element <"+child.getText()+">, is of type  <"+child.getNode().getElementType().toString()+">");
            if (child.getNode().getElementType() == iElementType) {
                //LOGGER.log(Level.INFO, "Child element <"+child.getText()+">is of token type: <"+iElementType.toString()+">");
                out.add(child);
            }
        }
        return out;
    }

    @Nullable
    public static ASTNode getNextNode(PsiElement compositeElement) {
        return compositeElement.getNode().getTreeNext();
    }

    @Nullable
    public static IElementType getNextNonEmptyNodeType(PsiElement compositeElement, boolean ignoreLineTerminator) {
        ASTNode next = getNextNonEmptyNode(compositeElement,ignoreLineTerminator);
        return next != null ? next.getElementType() : null;
    }

    @Nullable
    public static PsiElement getPreviousNonEmptySibling(PsiElement psiElement, boolean ignoreLineTerminator) {
        ASTNode node = getPreviousNonEmptyNode(psiElement, ignoreLineTerminator);
        return node != null ? node.getPsi() : null;
    }

    @Nullable
    public static PsiElement getNextNonEmptySibling(PsiElement psiElement, boolean ignoreLineTerminator) {
        ASTNode node = getNextNonEmptyNode(psiElement, ignoreLineTerminator);
        return node != null ? node.getPsi() : null;
    }

    @Nullable
    public static ASTNode getPreviousNonEmptyNode(PsiElement compositeElement, boolean ignoreLineTerminator) {
        ASTNode out = compositeElement != null ? compositeElement.getNode().getTreePrev() : null;
        while (shouldSkipNode(out, ignoreLineTerminator)) {
            if (out.getTreePrev() == null) {
                out = TreeUtil.prevLeaf(out);
            } else {
                out = out.getTreePrev();
            }
            if (out == null) {
                return null;
            }
            //LOGGER.log(Level.INFO, "<"+compositeElement.getText()+">NextNode "+out.getText()+" ElementType is <"+out.getElementType().toString()+">");
        }
        return out;
    }

    @Nullable
    public static ASTNode getNextNonEmptyNode(PsiElement compositeElement, boolean ignoreLineTerminator) {
        ASTNode out = compositeElement != null ? compositeElement.getNode().getTreeNext() : null;
        while (shouldSkipNode(out, ignoreLineTerminator)) {
            if (out.getTreeNext() == null) {
                out = TreeUtil.nextLeaf(out);
            } else {
                out = out.getTreeNext();
            }
            if (out == null) {
                return null;
            }
            //LOGGER.log(Level.INFO, "<"+compositeElement.getText()+">NextNode "+out.getText()+" ElementType is <"+out.getElementType().toString()+">");
        }
        return out;
    }

    public static boolean isWhitespaceOrEmpty(@Nullable PsiElement psiElement) {
        return psiElement == null || psiElement.getTextLength() == 0 || psiElement.getNode().getElementType() == TokenType.WHITE_SPACE;
    }

    public static boolean isWhitespaceOrEmpty(@Nullable ASTNode node) {
        return node == null || node.getTextLength() == 0 || node.getElementType() == TokenType.WHITE_SPACE;
    }

    private static boolean shouldSkipNode(ASTNode out, boolean ignoreLineTerminator) {
        return out != null && ((ignoreLineTerminator && out.getElementType() == ObjJTypes.ObjJ_LINE_TERMINATOR) || isWhitespaceOrEmpty(out) || out.getPsi() instanceof PsiErrorElement);
    }


    @Nullable
    public static <PsiT extends PsiElement> PsiT getSharedContextOfType(@Nullable PsiElement psiElement1, @Nullable PsiElement psiElement2, Class<PsiT> sharedClass) {
        if (psiElement1 == null || psiElement2 == null) {
            return null;
        }
        final PsiElement sharedContext = findCommonContext(psiElement1, psiElement2);
        if (sharedContext == null) {
            return null;
        }
        if (sharedClass.isInstance(sharedContext)) {
            return sharedClass.cast(sharedContext);
        }
        return getParentOfType(sharedContext, sharedClass);
    }

    public static <PsiT extends PsiElement> boolean hasSharedContextOfType(@Nullable PsiElement psiElement1, @Nullable PsiElement psiElement2, Class<PsiT> sharedClass) {
        return getSharedContextOfType(psiElement1, psiElement2, sharedClass) != null;
    }

    public static <PsiT extends PsiElement> boolean siblingOfTypeOccursAtLeastOnceBefore(@Nullable PsiElement psiElement, @NotNull Class<PsiT> siblingElementClass) {
        if (psiElement == null) {
            return false;
        }
        while (psiElement.getPrevSibling() != null) {
            psiElement = psiElement.getPrevSibling();
            if (siblingElementClass.isInstance(psiElement)) {
                return true;
            }
        }
        return false;
    }


    public static Spacing getLineBreak(Integer minLineFeeds,
                          boolean keepLineBreaks,
                          Integer keepBlankLines) {
        if (minLineFeeds == null) {
            minLineFeeds = 1;
        }
        if (keepBlankLines == null) {
            keepBlankLines = 1;
        }
        //return ObjJSpacingBuilder.createSpacingBuilder(0, Integer.MAX_VALUE, minLineFeeds, keepLineBreaks, keepBlankLines);

        return Spacing.createSafeSpacing(keepLineBreaks, keepBlankLines);
    }


    public static boolean hasLineBreakAfterInSameParent(@Nullable ASTNode node) {
        node = node != null ? node.getTreeNext() : null;
        return node != null &&  isWhitespaceWithLineBreak(TreeUtil.findLastLeaf(node));
    }

    public static boolean hasLineBreakBreakBeforeInSameParent(@Nullable ASTNode node) {
        node = node != null ? node.getTreePrev() : null;
        return node != null &&  isWhitespaceWithLineBreak(TreeUtil.findLastLeaf(node));
    }

    public static boolean isWhitespaceWithLineBreak(@Nullable ASTNode node) {
        return node != null && node.getElementType() == TokenType.WHITE_SPACE && node.textContains('\n');
    }

    public static boolean needsBlankLineBetweenItems(IElementType elementType1, IElementType elementType2) {
        if (COMMENTS.contains(elementType1) || COMMENTS.contains(elementType2)) {
            return false;
        }

        // Allow to keep consecutive runs of `use`, `const` or other "one line" items without blank lines
        if (elementType1 == elementType2 && ONE_LINE_ITEMS.contains(elementType1)) {
            return false;
        }

        return elementType1 != ObjJTypes.ObjJ_LINE_TERMINATOR;
    }

    @Nullable
    public static PsiElement getThisOrPreviousNonCommentElement(@Nullable PsiElement element) {
        if (element == null) {
            return null;
        }
        if (element instanceof ObjJComment) {
            return PsiTreeUtil.skipSiblingsBackward(element, PsiWhiteSpace.class, PsiComment.class);
        }
        return element;
    }


    @Nullable
    public static PsiElement getThisOrNextNonCommentElement(@Nullable PsiElement element) {
        if (element == null) {
            return null;
        }
        if (element instanceof ObjJComment) {
            return PsiTreeUtil.skipSiblingsForward(element, PsiWhiteSpace.class, PsiComment.class);
        }
        return element;
    }

}

File: src/org/cappuccino_project/ide/intellij/plugin/psi/utils/ObjJTreeUtil.kt
package org.cappuccino_project.ide.intellij.plugin.psi.utils

import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiErrorElement
import com.intellij.psi.impl.source.tree.TreeUtil
import com.intellij.psi.stubs.StubElement
import com.intellij.psi.tree.IElementType
import com.intellij.psi.util.PsiTreeUtil
import org.apache.velocity.runtime.parser.node.ASTMap
import org.cappuccino_project.ide.intellij.plugin.psi.types.ObjJTypes
import org.cappuccino_project.ide.intellij.plugin.utils.ArrayUtils

import java.util.ArrayList
import java.util.Collections
import java.util.logging.Level
import java.util.logging.Logger

class ObjJTreeUtil : PsiTreeUtil() {
    companion object {

        private val LOGGER = Logger.getLogger(ObjJTreeUtil::class.java.name)
        fun <StubT : StubElement<*>> filterStubChildren(parent: StubElement<com.intellij.psi.PsiElement>?, stubClass: Class<StubT>): List<StubT> {
            return if (parent == null) {
                emptyList()
            } else filterStubChildren(parent.childrenStubs, stubClass)
        }

        fun <StubT : StubElement<*>> filterStubChildren(children: List<StubElement<*>>?, stubClass: Class<StubT>): List<StubT> {
            return if (children == null) {
                emptyList()
            } else ArrayUtils.filter(children, stubClass)
        }

        fun getChildrenOfType(element: PsiElement, iElementType: IElementType): List<PsiElement> {
            val out = ArrayList<PsiElement>()
            for (child in element.children) {
                //LOGGER.log(Level.INFO, "Child element <"+child.getText()+">, is of type  <"+child.getNode().getElementType().toString()+">");
                if (child.node.elementType === iElementType) {
                    //LOGGER.log(Level.INFO, "Child element <"+child.getText()+">is of token type: <"+iElementType.toString()+">");
                    out.add(child)
                }
            }
            return out
        }

        fun getPreviousSiblingOfType(element: PsiElement, siblingElementType: IElementType): PsiElement? {
            var element = element
            while (element.prevSibling != null) {
                element = element.prevSibling
                if (hasElementType(element, siblingElementType)) {
                    return element
                }
            }
            return null
        }

        fun getNextSiblingOfType(element: PsiElement, siblingElementType: IElementType): PsiElement? {
            var element = element
            while (element.nextSibling != null) {
                element = element.nextSibling
                if (hasElementType(element, siblingElementType)) {
                    return element
                }
            }
            return null
        }

        private fun hasElementType(element: PsiElement, elementType: IElementType): Boolean {
            return element.node.elementType === elementType
        }

        fun getNextNodeType(compositeElement: PsiElement): IElementType? {
            val astNode = getNextNode(compositeElement)
            return astNode?.elementType
        }

        fun getNextNode(compositeElement: PsiElement): ASTNode? {
            return compositeElement.node.treeNext
        }

        fun getNextNonEmptyNodeType(compositeElement: PsiElement, ignoreLineTerminator: Boolean): IElementType? {
            val next = getNextNonEmptyNode(compositeElement, ignoreLineTerminator)
            return next?.elementType
        }

        fun getPreviousNonEmptySibling(psiElement: PsiElement, ignoreLineTerminator: Boolean): PsiElement? {
            val node = getPreviousNonEmptyNode(psiElement, ignoreLineTerminator)
            return node?.psi
        }

        fun getNextNonEmptySibling(psiElement: PsiElement, ignoreLineTerminator: Boolean): PsiElement? {
            val node = getNextNonEmptyNode(psiElement, ignoreLineTerminator)
            return node?.psi
        }

        fun getPreviousNonEmptyNode(compositeElement: PsiElement?, ignoreLineTerminator: Boolean): ASTNode? {
            var out: ASTNode? = compositeElement?.node?.treePrev
            while (shouldSkipNode(out, ignoreLineTerminator)) {
                if (out!!.treePrev == null) {
                    out = TreeUtil.prevLeaf(out)
                } else {
                    out = out.treePrev
                }
                if (out == null) {
                    return null
                }
                //LOGGER.log(Level.INFO, "<"+compositeElement.getText()+">NextNode "+out.getText()+" ElementType is <"+out.getElementType().toString()+">");
            }
            return out
        }

        fun getNextNonEmptyNode(compositeElement: PsiElement?, ignoreLineTerminator: Boolean): ASTNode? {
            var out: ASTNode? = compositeElement?.node?.treeNext
            while (shouldSkipNode(out, ignoreLineTerminator)) {
                if (out!!.treeNext == null) {
                    out = TreeUtil.nextLeaf(out)
                } else {
                    out = out.treeNext
                }
                if (out == null) {
                    return null
                }
                //LOGGER.log(Level.INFO, "<"+compositeElement.getText()+">NextNode "+out.getText()+" ElementType is <"+out.getElementType().toString()+">");
            }
            return out
        }

        private fun shouldSkipNode(out: ASTNode?, ignoreLineTerminator: Boolean): Boolean {
            return out != null && (ignoreLineTerminator && out.elementType === ObjJTypes.ObjJ_LINE_TERMINATOR || out.elementType === com.intellij.psi.TokenType.WHITE_SPACE || out.psi is PsiErrorElement)
        }


        fun <PsiT : PsiElement> getSharedContextOfType(psiElement1: PsiElement?, psiElement2: PsiElement?, sharedClass: Class<PsiT>): PsiT? {
            if (psiElement1 == null || psiElement2 == null) {
                return null
            }
            val sharedContext = PsiTreeUtil.findCommonContext(psiElement1, psiElement2) ?: return null
            return if (sharedClass.isInstance(sharedContext)) {
                sharedClass.cast(sharedContext)
            } else PsiTreeUtil.getParentOfType(sharedContext, sharedClass)
        }

        fun <PsiT : PsiElement> hasSharedContextOfType(psiElement1: PsiElement?, psiElement2: PsiElement?, sharedClass: Class<PsiT>): Boolean {
            return getSharedContextOfType(psiElement1, psiElement2, sharedClass) != null
        }

        fun <PsiT : PsiElement> siblingOfTypeOccursAtLeastOnceBefore(psiElement: PsiElement?, siblingElementClass: Class<PsiT>): Boolean {
            var psiElement: PsiElement? = psiElement ?: return false
            while (psiElement!!.prevSibling != null) {
                psiElement = psiElement.prevSibling
                if (siblingElementClass.isInstance(psiElement)) {
                    return true
                }
            }
            return false
        }
    }

}
